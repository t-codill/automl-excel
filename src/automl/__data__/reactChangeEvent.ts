// tslint:disable: no-any
export const reactChangeEvent: React.ChangeEvent<any> = {
    target: {}
} as unknown as React.ChangeEvent<any>;
