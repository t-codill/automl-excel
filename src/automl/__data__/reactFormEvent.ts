// tslint:disable: no-any
export const reactFormEvent: React.FormEvent<any> = {
    preventDefault(): void { return; }
} as unknown as React.FormEvent<any>;
