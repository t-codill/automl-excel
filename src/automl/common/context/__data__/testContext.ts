import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
