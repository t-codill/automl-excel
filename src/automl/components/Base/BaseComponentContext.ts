import React from "react";
import { WorkspaceContext } from "../../common/context/WorkspaceContext";

export abstract class BaseComponentContext<TProp, TState> extends React.PureComponent<TProp, TState> {
    public static contextType = WorkspaceContext;
    public context!: React.ContextType<typeof WorkspaceContext>;
}
