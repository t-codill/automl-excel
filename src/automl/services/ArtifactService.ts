import { ArtifactAPI, ArtifactAPIModels } from "@vienna/artifact";
import { blob2string } from "../common/utils/blob2string";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactClientApiVersion = "1.0";
const runOrigin = "ExperimentRun";
const modelPath = "outputs/model.pkl";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export interface IRunDataForArtifact {
    runId: string | undefined;
    parentRunId: string | undefined;
    status: string | undefined;
}

async function getArtifactContent(artifact: ArtifactAPIModels.GetArtifactContentByIdResponse): Promise<string | undefined> {
    const blobBody = await artifact.blobBody;
    if (blobBody) {
        return blob2string(blobBody);
    }
    return undefined;
}
export class ArtifactService extends ServiceBaseNonArm<ArtifactAPI> {
    constructor(props: IServiceBaseProps) {
        super(props, ArtifactAPI, props.discoverUrls.history);
    }
    public async tryGetContentForRun(runId: string, path: string): Promise<string | null | undefined> {
        return this.tryGetContent({ origin: runOrigin, container: runId, path });
    }

    public async tryGetContent(artifact: IArtifact): Promise<string | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getContent(artifact: IArtifact): Promise<string | undefined> {
        return this.send(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return this.parallelGetAllValues(params, async (param, client, abortSignal) => {
            const artifact = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                param.origin,
                param.container,
                artifactClientApiVersion, {
                    path: param.path,
                    abortSignal
                });
            return getArtifactContent(artifact);
        });
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return this.parallelGetAllValuesWithContinuationToken(runs, async (runId, client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async getAllArtifacts(runId: string): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async tryGetArtifactUrl(artifact: IArtifact): Promise<ArtifactAPIModels.GetArtifactContentInformation2Response | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            return client.getArtifactContentInformation2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    abortSignal,
                    path: artifact.path
                });
        });
    }

    public async getModelUrl(run: IRunDataForArtifact): Promise<string | null | undefined> {
        if (!run || !run.runId || !run.parentRunId || run.status !== "Completed") {
            return null;
        }
        const result = await this.tryGetArtifactUrl({ container: run.runId, origin: runOrigin, path: modelPath });
        if (result === undefined) {
            return undefined;
        }
        if (!result || !result.contentUri) {
            return null;
        }
        return result.contentUri;
    }
}
