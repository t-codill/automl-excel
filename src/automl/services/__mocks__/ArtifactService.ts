import { ArtifactAPIModels } from "@vienna/artifact";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export class ArtifactService {
    private readonly tryGetContentResult = "tryGetContent";
    private readonly getAllArtifactsResult: ArtifactAPIModels.ArtifactDto[] = [
        {
            artifactId: "artifactId1",
            container: "container1",
            origin: "origin1",
            path: "path1"
        },
        {
            artifactId: "artifactId2",
            container: "container2",
            origin: "origin2",
            path: "path2"
        }
    ];
    private readonly tryGetArtifactUrlResult: ArtifactAPIModels.ArtifactContentInformationDto = {
        contentUri: "contentUri",
        path: "path"
    };
    private readonly getModelUrlResult = "getModelUrl";
    private readonly getModelUrlsResult = ["getModelUrls1", "getModelUrls2"];

    public async tryGetContentForRun(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async tryGetContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return params.map((p) => `"getAllContentsResult:${p.origin}/${p.container}/${p.path}`);
    }

    public async getAllArtifacts(): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllArtifactsResult;
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return runs.map(() => this.getAllArtifactsResult);
    }

    public async tryGetArtifactUrl(): Promise<ArtifactAPIModels.ArtifactContentInformationDto | null | undefined> {
        return this.tryGetArtifactUrlResult;
    }

    public async getModelUrl(): Promise<string | undefined> {
        return this.getModelUrlResult;
    }

    public async getModelUrls(): Promise<Array<string | null> | undefined> {
        return this.getModelUrlsResult;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
