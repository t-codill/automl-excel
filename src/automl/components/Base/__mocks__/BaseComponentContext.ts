import React from "react";
import { testContext } from "../../../common/context/__data__/testContext";
import { IWorkspaceContextProps } from "../../../common/context/IWorkspaceProps";
import { WorkspaceContext } from "../../../common/context/WorkspaceContext";

export abstract class BaseComponentContext<TProp, TState> extends React.Component<TProp, TState> {
    public static contextType = WorkspaceContext;
    public set context(_value: IWorkspaceContextProps) {
        return;
    }
    public get context(): React.ContextType<typeof WorkspaceContext> {
        return testContext;
    }
}
