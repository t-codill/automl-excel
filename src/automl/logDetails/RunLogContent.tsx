import * as React from "react";
import MonacoEditor from "react-monaco-editor";
import { BaseComponent } from "../components/Base/BaseComponent";
import { PageLoadingSpinner } from "../components/Progress/PageLoadingSpinner";
import { ArtifactService } from "../services/ArtifactService";
import { IRunLog } from "./RunLogs";

export interface IRunLogProps {
    runLog: IRunLog;
    refreshTimestamp?: number;
}
export class RunLogContent extends BaseComponent<IRunLogProps,
    { content: string | undefined },
    { artifactService: ArtifactService }> {
    protected readonly header = "Run Detail";
    protected readonly serviceConstructors = { artifactService: ArtifactService };

    public render(): React.ReactNode {
        if (!this.state || !this.state.content) {
            return <PageLoadingSpinner />;
        }
        const options = {
            readOnly: true,
            automaticLayout: true
        };

        return <MonacoEditor
            language="text"
            theme={this.context.theme === "dark" ? "vs-dark" : "vs"}
            value={this.state.content}
            options={options}
        />;
    }

    public componentDidUpdate(prevProps: IRunLogProps): void {
        if (prevProps.refreshTimestamp !== this.props.refreshTimestamp) {
            this.refresh();
        }
    }
    protected getData = async () => {
        const content = await this.services.artifactService.getContent(this.props.runLog);
        if (!content) {
            return;
        }

        this.setState({ content });
    }
}
