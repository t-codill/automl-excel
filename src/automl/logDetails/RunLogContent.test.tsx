import { shallow } from "enzyme";
import * as React from "react";
import { testContext } from "../common/context/__data__/testContext";
import { BaseComponentContext } from "../components/Base/BaseComponentContext";
import { ArtifactService } from "../services/ArtifactService";
import { RunLogContent } from "./RunLogContent";

jest.mock("../services/ArtifactService");
let asGetContent: jest.SpyInstance<ReturnType<ArtifactService["getContent"]>>;
let themeMock: jest.SpyInstance<BaseComponentContext<{}, {}>["context"]>;

describe("Run Log Content Loading", () => {
    const log = {
        logName: "log1",
        prefix: "l1",
        origin: "origin",
        container: "container",
        path: "path"
    };
    beforeEach(() => {
        asGetContent = jest.spyOn(ArtifactService.prototype, "getContent");
        themeMock = jest.spyOn(BaseComponentContext.prototype, "context", "get");
    });
    it("should render spinner when no content", async () => {
        const getContentMock = Promise.resolve(undefined);
        asGetContent.mockReturnValueOnce(getContentMock);
        const tree = shallow(<RunLogContent runLog={log} />);
        await getContentMock;
        expect(tree)
            .toMatchInlineSnapshot("<pageLoadingSpinner />");
    });
    it("should try get data", async () => {
        asGetContent.mockReturnValueOnce(Promise.resolve(undefined));
        const tree = shallow<RunLogContent>(<RunLogContent runLog={log} />);
        await Promise.resolve();
        expect(tree)
            .toMatchInlineSnapshot("<pageLoadingSpinner />");
        asGetContent.mockClear();
        tree.setProps({
            refreshTimestamp: 1
        });
        await Promise.resolve();
        expect(asGetContent)
            .toBeCalledTimes(1);
    });
});
describe("Run Log Context Theme", () => {
    const log = {
        logName: "log1",
        prefix: "l1",
        origin: "origin",
        container: "container",
        path: "path"
    };
    beforeEach(() => {
        themeMock = jest.spyOn(BaseComponentContext.prototype, "context", "get");
    });
    it("should render log content when valid content with light theme", async () => {
        const getContentMock = Promise.resolve("light theme content");
        asGetContent.mockReturnValueOnce(getContentMock);
        const tree = shallow(<RunLogContent runLog={log} />);
        await getContentMock;
        expect(tree)
            .toMatchSnapshot();
    });
    it("should render log content when dark theme", async () => {
        const getContentMock = Promise.resolve("dark theme content");
        asGetContent.mockReturnValueOnce(getContentMock);
        themeMock.mockImplementation(() => {
            const newContext = { ...testContext, theme: "dark" };
            return newContext;
        });
        const tree = shallow(<RunLogContent runLog={log} />);
        await getContentMock;
        expect(tree)
            .toMatchSnapshot();
    });
});
