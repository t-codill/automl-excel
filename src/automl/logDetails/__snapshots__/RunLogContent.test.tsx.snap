// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run Log Context Theme should render log content when dark theme 1`] = `
<MonacoEditor
  language="text"
  options={
    Object {
      "automaticLayout": true,
      "readOnly": true,
    }
  }
  theme="vs-dark"
  value="tryGetContent"
/>
`;

exports[`Run Log Context Theme should render log content when valid content with light theme 1`] = `
<MonacoEditor
  language="text"
  options={
    Object {
      "automaticLayout": true,
      "readOnly": true,
    }
  }
  theme="vs"
  value="tryGetContent"
/>
`;
